src/whylogs/core/statistics/constraints.py

```diff
@@ -0,0 +1,272 @@
+            raise ValueError("Summary constraint must specify a second value or field name, but not both")
+
+    @property
+    def name(self):
+        return self._name if self._name is not None else f'summary {self.field} {Op.Name(self.op)} {self.value}/{self.field2}'
+
+    def track(self, summ: NumberSummary) -> bool:
+        self.total += 1
+        if not self.func(summ):
+            self.failures += 1
+            if self._verbose:
+                logger.info(f'summary constraint {self.name} failed')
+
+    @staticmethod
+    def from_protobuf(msg: SummaryConstraintMsg) -> 'SummaryConstraint':
+        return ValueConstraint(msg.op, msg.value, name=msg.name)
+
+    def to_protobuf(self) -> SummaryConstraintMsg:
+        if self.field2 is None:
+            msg = SummaryConstraintMsg(
+                name=self.name,
+                field=self.field,
+                op=self.op,
+                value=self.value,
+            )
+        else:
+            msg = SummaryConstraintMsg(
+                name=self.name,
+                field=self.field,
+                op=self.op,
+                field2=self.field2,
+            )
+        return msg
+
+    def report(self):
+        return (self.name, self.total, self.failures)
+
+
+class ValueConstraints:
+    def __init__(self, constraints: List[ValueConstraint] = []):
+        self.constraints = constraints
+
+    @staticmethod
+    def from_protobuf(msg: ValueConstraintMsgs) -> 'ValueConstraints':
+        v = [ValueConstraint.from_proto(c) for c in msg.constraints]
+        if len(v) > 0:
+            return ValueConstraints(v)
+        return None
+
+    def to_protobuf(self) -> ValueConstraintMsgs:
+        v = [c.to_protobuf() for c in self.constraints]
+        if len(v) > 0:
+            vcmsg = ValueConstraintMsgs()
+            vcmsg.constraints.extend(v)
+            return vcmsg
+        return None
+
+    def add_constraint(self, c: ValueConstraint):
+        self.constraints.append(c)
+
+    def track(self, v):
+        for c in self.constraints:
+            c.track(v)
+
+    def report(self) -> List[tuple]:
+        v = [c.report() for c in self.constraints]
+        if len(v) > 0:
+            return v
+        return None
+
+
+class SummaryConstraints:
+    def __init__(self, constraints: List[SummaryConstraint]):
+        self.constraints = constraints
+
+    @staticmethod
+    def from_protobuf(msg: SummaryConstraintMsgs) -> 'SummaryConstraints':
+        v = [SummaryConstraint.from_proto(c) for c in msg.constraints]
+        if len(v) > 0:
+            return SummaryConstraints(v)
+        return None
+
+    def to_protobuf(self) -> SummaryConstraintMsgs:
+        v = [c.to_protobuf() for c in self.constraints]
+        if len(v) > 0:
+            scmsg = SummaryConstraintMsgs()
+            scmsg.constraints.extend(v)
+            return scmsg
+        return None
+
+    def add_constraint(self, c: SummaryConstraint):
+        self.constraints.append(c)
```

Comment:
What's the use case of adding a new constraint after creation? Should we modify the existing object or return a new `SummaryConstraints` here? I would be cautious about state modification and would prefer returning a new object in this case

Reply:
Noe use-case, I was anticipating a non-existent need.  Removed this method.